src/controllers/krate/versions.rs

```diff
@@ -21,27 +27,217 @@ pub async fn versions(state: AppState, Path(crate_name): Path<String>) -> AppRes
+        if params.get("per_page").is_some() {
+            pagination = Some(
+                PaginationOptions::builder()
+                    .enable_seek(true)
+                    .enable_pages(false)
+                    .gather(&req)?,
+            );
+        }
 
-        versions_and_publishers
-            .sort_by_cached_key(|(version, _)| Reverse(semver::Version::parse(&version.num).ok()));
+        // Sort by semver by default
+        let versions_and_publishers = match params.get("sort").map(|s| s.to_lowercase()).as_deref()
+        {
+            Some("date") => list_by_date(&krate, pagination.as_ref(), &req, conn)?,
+            _ => list_by_semver(&krate, pagination.as_ref(), &req, conn)?,
+        };
 
         let versions = versions_and_publishers
+            .data
             .iter()
             .map(|(v, _)| v)
             .cloned()
             .collect::<Vec<_>>();
         let versions = versions_and_publishers
+            .data
             .into_iter()
             .zip(VersionOwnerAction::for_versions(conn, &versions)?)
             .map(|((v, pb), aas)| EncodableVersion::from(v, &crate_name, pb, aas))
             .collect::<Vec<_>>();
 
-        Ok(Json(json!({ "versions": versions })))
+        Ok(Json(match pagination {
+            Some(_) => json!({ "versions": versions, "meta": versions_and_publishers.meta }),
+            None => json!({ "versions": versions }),
+        }))
     })
     .await
 }
+
+fn list_by_date(
+    krate: &Crate,
+    options: Option<&PaginationOptions>,
+    req: &Parts,
+    conn: &mut PgConnection,
+) -> AppResult<PaginatedVersionsAndPublishers> {
+    let mut query = krate
+        .all_versions()
+        .left_outer_join(users::table)
+        .select((versions::all_columns, users::all_columns.nullable()));
+
+    if let Some(options) = options {
+        if let Page::Seek(ref seek) = options.page {
+            let (created_at, id) = seek.decode::<seek::Date>().map(|s| (s.0, s.1))?;
+            query = query.filter(
+                versions::created_at
+                    .eq(created_at)
+                    .and(versions::id.lt(id))
+                    .or(versions::created_at.lt(created_at)),
+            )
+        }
+        query = query.limit(options.per_page);
+    }
+
+    query = query.order((versions::created_at.desc(), versions::id.desc()));
+
+    let data: Vec<(Version, Option<User>)> = query.load(conn)?;
+    let mut next_page = None;
+    if let Some(options) = options {
+        next_page = next_seek_params(&data, options, |last| {
+            seek::Date(last.0.created_at, last.0.id)
+        })?
+        .map(|p| req.query_with_params(p));
+    };
+
+    // Since the total count is retrieved through an additional query, to maintain consistency
+    // with other pagination methods, we only make a count query while data is not empty.
+    let total = if !data.is_empty() {
+        versions::table.count().get_result(conn)?
```

Comment:
I think there is a bug here, since this will return the number of all versions in the database, not just the ones belonging to this crate. I assume this would have to be something like `krate.all_versions().count()`.

Reply:
Good catch! I should also make it fail in the test suite.